# Abstract



1. This analysis is part of kaggle competition which has been given as assignment for INFO6105 class by out Professor Nick Bear Brown. In this we have a dataset of Inter American development bank for households of costa Rican families poverty levels. We will be assessing the levels of Poverty namely VULNERABLE HOUSEHOLDS, NON VULNERABLE HOUSEHOLDS, MODERATE POVERTY, EXTREME POVERTY based on factors like household goods, condition of house, education of people, family size etc. We have succesfully classified the houses with a macro f1 score of 0.41064.

Details of column names:

v2a1, Monthly rent payment

hacdor, =1 Overcrowding by bedrooms

rooms,  number of all rooms in the house

hacapo, =1 Overcrowding by rooms

v14a, =1 has bathroom in the household

refrig, =1 if the household has refrigerator

v18q, owns a tablet

v18q1, number of tablets household owns

r4h1, Males younger than 12 years of age

r4h2, Males 12 years of age and older

r4h3, Total males in the household

r4m1, Females younger than 12 years of age

r4m2, Females 12 years of age and older

r4m3, Total females in the household

r4t1, persons younger than 12 years of age

r4t2, persons 12 years of age and older

r4t3, Total persons in the household

tamhog, size of the household

tamviv, number of persons living in the household

escolari, years of schooling

rez_esc, Years behind in school

hhsize, household size

paredblolad, =1 if predominant material on the outside wall is block or brick

paredzocalo, "=1 if predominant material on the outside wall is socket (wood,  zinc or absbesto"

paredpreb, =1 if predominant material on the outside wall is prefabricated or cement

pareddes, =1 if predominant material on the outside wall is waste material

paredmad, =1 if predominant material on the outside wall is wood

paredzinc, =1 if predominant material on the outside wall is zink

paredfibras, =1 if predominant material on the outside wall is natural fibers

paredother, =1 if predominant material on the outside wall is other

pisomoscer, "=1 if predominant material on the floor is mosaic,  ceramic,  terrazo"

pisocemento, =1 if predominant material on the floor is cement

pisoother, =1 if predominant material on the floor is other

pisonatur, =1 if predominant material on the floor is  natural material

pisonotiene, =1 if no floor at the household

pisomadera, =1 if predominant material on the floor is wood

techozinc, =1 if predominant material on the roof is metal foil or zink

techoentrepiso, "=1 if predominant material on the roof is fiber cement,  mezzanine "

techocane, =1 if predominant material on the roof is natural fibers

techootro, =1 if predominant material on the roof is other

cielorazo, =1 if the house has ceiling

abastaguadentro, =1 if water provision inside the dwelling

abastaguafuera, =1 if water provision outside the dwelling

abastaguano, =1 if no water provision

public, "=1 electricity from CNFL,  ICE,  ESPH/JASEC"

planpri, =1 electricity from private plant

noelec, =1 no electricity in the dwelling

coopele, =1 electricity from cooperative

sanitario1, =1 no toilet in the dwelling

sanitario2, =1 toilet connected to sewer or cesspool

sanitario3, =1 toilet connected to  septic tank

sanitario5, =1 toilet connected to black hole or letrine

sanitario6, =1 toilet connected to other system

energcocinar1, =1 no main source of energy used for cooking (no kitchen)

energcocinar2, =1 main source of energy used for cooking electricity

energcocinar3, =1 main source of energy used for cooking gas

energcocinar4, =1 main source of energy used for cooking wood charcoal

elimbasu1, =1 if rubbish disposal mainly by tanker truck

elimbasu2, =1 if rubbish disposal mainly by botan hollow or buried

elimbasu3, =1 if rubbish disposal mainly by burning

elimbasu4, =1 if rubbish disposal mainly by throwing in an unoccupied space

elimbasu5, "=1 if rubbish disposal mainly by throwing in river,  creek or sea"

elimbasu6, =1 if rubbish disposal mainly other

epared1, =1 if walls are bad

epared2, =1 if walls are regular

epared3, =1 if walls are good

etecho1, =1 if roof are bad

etecho2, =1 if roof are regular

etecho3, =1 if roof are good

eviv1, =1 if floor are bad

eviv2, =1 if floor are regular

eviv3, =1 if floor are good

dis, =1 if disable person

male, =1 if male

female, =1 if female

estadocivil1, =1 if less than 10 years old

estadocivil2, =1 if free or coupled uunion

estadocivil3, =1 if married

estadocivil4, =1 if divorced

estadocivil5, =1 if separated

estadocivil6, =1 if widow/er

estadocivil7, =1 if single

parentesco1, =1 if household head

parentesco2, =1 if spouse/partner

parentesco3, =1 if son/doughter

parentesco4, =1 if stepson/doughter

parentesco5, =1 if son/doughter in law

parentesco6, =1 if grandson/doughter

parentesco7, =1 if mother/father

parentesco8, =1 if father/mother in law

parentesco9, =1 if brother/sister

parentesco10, =1 if brother/sister in law

parentesco11, =1 if other family member

parentesco12, =1 if other non family member

idhogar, Household level identifier

hogar_nin, Number of children 0 to 19 in household

hogar_adul, Number of adults in household

hogar_mayor, # of individuals 65+ in the household

hogar_total, # of total individuals in the household

dependency, Dependency rate, calculated = (number of members of the household younger than 19 or older than 64)/(number of member of household between 19 and 64)

edjefe, years of education of male head of household, based on the interaction of escolari (years of education), head of household and gender, yes=1 and no=0

edjefa, years of education of female head of household, based on the interaction of escolari (years of education), head of household and gender, yes=1 and no=0

meaneduc,average years of education for adults (18+)

instlevel1, =1 no level of education

instlevel2, =1 incomplete primary

instlevel3, =1 complete primary

instlevel4, =1 incomplete academic secondary level

instlevel5, =1 complete academic secondary level

instlevel6, =1 incomplete technical secondary level

instlevel7, =1 complete technical secondary level

instlevel8, =1 undergraduate and higher education

instlevel9, =1 postgraduate higher education


bedrooms, number of bedrooms

overcrowding, # persons per room

tipovivi1, =1 own and fully paid house

tipovivi2, "=1 own,  paying in installments"

tipovivi3, =1 rented

tipovivi4, =1 precarious

tipovivi5, "=1 other(assigned,  borrowed)"

computer, =1 if the household has notebook or desktop computer

television, =1 if the household has TV

mobilephone, =1 if mobile phone

qmobilephone, # of mobile phones

lugar1, =1 region Central

lugar2, =1 region Chorotega

lugar3, =1 region Pacífico central

lugar4, =1 region Brunca

lugar5, =1 region Huetar Atlántica

lugar6, =1 region Huetar Norte

area1, =1 zona urbana

area2, =2 zona rural

age, Age in years

SQBescolari, escolari squared

SQBage, age squared

SQBhogar_total, hogar_total squared

SQBedjefe, edjefe squared

SQBhogar_nin, hogar_nin squared

SQBovercrowding, overcrowding squared

SQBdependency, dependency squared

SQBmeaned, square of the mean years of education of adults (>=18) in the household

agesq, Age squared

In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load in 

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the "../input/" directory.
# For example, running this (by clicking run or pressing Shift+Enter) will list the files in the input directory

import os
print(os.listdir("../input"))

# Any results you write to the current directory are saved as output.

['test.csv', 'train.csv', 'sample_submission.csv']


In [2]:
import numpy as np
import pandas as pd
import os
from plotly.offline import init_notebook_mode, iplot
import plotly.graph_objs as go
import plotly.plotly as py
from plotly import tools
from datetime import date
import pandas as pd
import numpy as np 
import plotly.figure_factory as ff

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
import lightgbm as lgb

import matplotlib.pyplot as plt
import seaborn as sns
import random 
import warnings
import operator
warnings.filterwarnings("ignore")
init_notebook_mode(connected=True)

In [3]:
train=pd.read_csv('../input/train.csv')
test=pd.read_csv('../input/test.csv')

In [4]:
test['edjefe'].dtypes

dtype('O')

'edjefe' is years of education of head of household when male.

In [5]:
train.columns[train.dtypes==object]

Index(['Id', 'idhogar', 'dependency', 'edjefe', 'edjefa'], dtype='object')

In [6]:
train['dependency'].unique()

array(['no', '8', 'yes', '3', '.5', '.25', '2', '.66666669', '.33333334',
       '1.5', '.40000001', '.75', '1.25', '.2', '2.5', '1.2', '4',
       '1.3333334', '2.25', '.22222222', '5', '.83333331', '.80000001',
       '6', '3.5', '1.6666666', '.2857143', '1.75', '.71428573',
       '.16666667', '.60000002'], dtype=object)

In [7]:
train[(train['dependency']=='no') & (train['SQBdependency']!=0)]

Empty DataFrame
Columns: [Id, v2a1, hacdor, rooms, hacapo, v14a, refrig, v18q, v18q1, r4h1, r4h2, r4h3, r4m1, r4m2, r4m3, r4t1, r4t2, r4t3, tamhog, tamviv, escolari, rez_esc, hhsize, paredblolad, paredzocalo, paredpreb, pareddes, paredmad, paredzinc, paredfibras, paredother, pisomoscer, pisocemento, pisoother, pisonatur, pisonotiene, pisomadera, techozinc, techoentrepiso, techocane, techootro, cielorazo, abastaguadentro, abastaguafuera, abastaguano, public, planpri, noelec, coopele, sanitario1, sanitario2, sanitario3, sanitario5, sanitario6, energcocinar1, energcocinar2, energcocinar3, energcocinar4, elimbasu1, elimbasu2, elimbasu3, elimbasu4, elimbasu5, elimbasu6, epared1, epared2, epared3, etecho1, etecho2, etecho3, eviv1, eviv2, eviv3, dis, male, female, estadocivil1, estadocivil2, estadocivil3, estadocivil4, estadocivil5, estadocivil6, estadocivil7, parentesco1, parentesco2, parentesco3, parentesco4, parentesco5, parentesco6, parentesco7, parentesco8, parentesco9, parentesco10, parentesco11, parentesco12, idhogar, hogar_nin, hogar_adul, hogar_mayor, hogar_total, ...]
Index: []

[0 rows x 143 columns]

So square of no is zero

In [8]:
train[(train['dependency']=='yes') & (train['SQBdependency']!=1)]

Empty DataFrame
Columns: [Id, v2a1, hacdor, rooms, hacapo, v14a, refrig, v18q, v18q1, r4h1, r4h2, r4h3, r4m1, r4m2, r4m3, r4t1, r4t2, r4t3, tamhog, tamviv, escolari, rez_esc, hhsize, paredblolad, paredzocalo, paredpreb, pareddes, paredmad, paredzinc, paredfibras, paredother, pisomoscer, pisocemento, pisoother, pisonatur, pisonotiene, pisomadera, techozinc, techoentrepiso, techocane, techootro, cielorazo, abastaguadentro, abastaguafuera, abastaguano, public, planpri, noelec, coopele, sanitario1, sanitario2, sanitario3, sanitario5, sanitario6, energcocinar1, energcocinar2, energcocinar3, energcocinar4, elimbasu1, elimbasu2, elimbasu3, elimbasu4, elimbasu5, elimbasu6, epared1, epared2, epared3, etecho1, etecho2, etecho3, eviv1, eviv2, eviv3, dis, male, female, estadocivil1, estadocivil2, estadocivil3, estadocivil4, estadocivil5, estadocivil6, estadocivil7, parentesco1, parentesco2, parentesco3, parentesco4, parentesco5, parentesco6, parentesco7, parentesco8, parentesco9, parentesco10, parentesco11, parentesco12, idhogar, hogar_nin, hogar_adul, hogar_mayor, hogar_total, ...]
Index: []

[0 rows x 143 columns]

square of yes is One

Therefore we can derive following code:

In [9]:
train['dependency']=np.sqrt(train['SQBdependency'])

In [10]:
train['edjefe'].unique()

array(['10', '12', 'no', '11', '9', '15', '4', '6', '8', '17', '7', '16',
       '14', '5', '21', '2', '19', 'yes', '3', '18', '13', '20'],
      dtype=object)

In [11]:
train['edjefa'].unique()

array(['no', '11', '4', '10', '9', '15', '7', '14', '13', '8', '17', '6',
       '5', '3', '16', '19', 'yes', '21', '12', '2', '20', '18'],
      dtype=object)

In [12]:
train['SQBedjefe'].unique()

array([100, 144,   0, 121,  81, 225,  16,  36,  64, 289,  49, 256, 196,
        25, 441,   4, 361,   1,   9, 324, 169, 400])

In [13]:
train[['edjefe', 'edjefa', 'SQBedjefe']][:20]

edjefe edjefa  SQBedjefe
0      10     no        100
1      12     no        144
2      no     11          0
3      11     no        121
4      11     no        121
5      11     no        121
6      11     no        121
7       9     no         81
8       9     no         81
9       9     no         81
10      9     no         81
11     no     11          0
12     no     11          0
13     no      4          0
14     no      4          0
15     no     10          0
16     no     10          0
17     no     10          0
18     no     10          0
19     15     no        225

In [14]:
train[['edjefe', 'edjefa', 'SQBedjefe']][train['edjefe']=='yes']

edjefe edjefa  SQBedjefe
601     yes     no          1
602     yes     no          1
2392    yes     no          1
2393    yes     no          1
2394    yes     no          1
2422    yes     no          1
2423    yes     no          1
2424    yes     no          1
2829    yes     no          1
2830    yes     no          1
2831    yes     no          1
3015    yes     no          1
3016    yes     no          1
3017    yes     no          1
3741    yes     no          1
3742    yes     no          1
3743    yes     no          1
3744    yes     no          1
3745    yes     no          1
4490    yes     no          1
4491    yes     no          1
4492    yes     no          1
4493    yes     no          1
4494    yes     no          1
4903    yes     no          1
4904    yes     no          1
4905    yes     no          1
4979    yes     no          1
4980    yes     no          1
4981    yes     no          1
...     ...    ...        ...
8385    yes     no          1
8585    yes     no          1
8903    yes     no          1
8904    yes     no          1
8905    yes     no          1
8951    yes     no          1
8952    yes     no          1
8953    yes     no          1
8954    yes     no          1
8955    yes     no          1
8956    yes     no          1
8957    yes     no          1
9040    yes     no          1
9041    yes     no          1
9042    yes     no          1
9043    yes     no          1
9044    yes     no          1
9045    yes     no          1
9046    yes     no          1
9047    yes     no          1
9197    yes     no          1
9198    yes     no          1
9199    yes     no          1
9257    yes     no          1
9415    yes     no          1
9416    yes     no          1
9417    yes     no          1
9463    yes     no          1
9464    yes     no          1
9465    yes     no          1

[123 rows x 3 columns]

In [15]:
train[(train['edjefe']=='yes') & (train['edjefa']!='no')]

Empty DataFrame
Columns: [Id, v2a1, hacdor, rooms, hacapo, v14a, refrig, v18q, v18q1, r4h1, r4h2, r4h3, r4m1, r4m2, r4m3, r4t1, r4t2, r4t3, tamhog, tamviv, escolari, rez_esc, hhsize, paredblolad, paredzocalo, paredpreb, pareddes, paredmad, paredzinc, paredfibras, paredother, pisomoscer, pisocemento, pisoother, pisonatur, pisonotiene, pisomadera, techozinc, techoentrepiso, techocane, techootro, cielorazo, abastaguadentro, abastaguafuera, abastaguano, public, planpri, noelec, coopele, sanitario1, sanitario2, sanitario3, sanitario5, sanitario6, energcocinar1, energcocinar2, energcocinar3, energcocinar4, elimbasu1, elimbasu2, elimbasu3, elimbasu4, elimbasu5, elimbasu6, epared1, epared2, epared3, etecho1, etecho2, etecho3, eviv1, eviv2, eviv3, dis, male, female, estadocivil1, estadocivil2, estadocivil3, estadocivil4, estadocivil5, estadocivil6, estadocivil7, parentesco1, parentesco2, parentesco3, parentesco4, parentesco5, parentesco6, parentesco7, parentesco8, parentesco9, parentesco10, parentesco11, parentesco12, idhogar, hogar_nin, hogar_adul, hogar_mayor, hogar_total, ...]
Index: []

[0 rows x 143 columns]

In [16]:
train[(train['edjefa']=='yes') & (train['parentesco1']==1)][['edjefe', 'edjefa', 'parentesco1', 'escolari']]

edjefe edjefa  parentesco1  escolari
432      no    yes            1         1
568      no    yes            1         1
595      no    yes            1         1
976      no    yes            1         1
1574     no    yes            1         1
2464     no    yes            1         1
2614     no    yes            1         1
3469     no    yes            1         1
3855     no    yes            1         1
3952     no    yes            1         1
4197     no    yes            1         1
4487     no    yes            1         1
4867     no    yes            1         1
4888     no    yes            1         1
4920     no    yes            1         1
6030     no    yes            1         1
6067     no    yes            1         1
6427     no    yes            1         1
7117     no    yes            1         1
7887     no    yes            1         1
8134     no    yes            1         1
8164     no    yes            1         1
8437     no    yes            1         1
8509     no    yes            1         1
9166     no    yes            1         1

This data shows when head is male and he is the head of the house the years of schooling is one.

In [17]:
train[train['edjefe']=='yes'][['edjefe', 'edjefa','age', 'escolari', 'parentesco1','male', 'female', 'idhogar']]

edjefe edjefa  age    ...      male  female    idhogar
601     yes     no   81    ...         1       0  3641ce2d1
602     yes     no   73    ...         0       1  3641ce2d1
2392    yes     no   42    ...         0       1  a7abf59cc
2393    yes     no   12    ...         0       1  a7abf59cc
2394    yes     no   46    ...         1       0  a7abf59cc
2422    yes     no   41    ...         1       0  698318dd7
2423    yes     no   74    ...         0       1  698318dd7
2424    yes     no   78    ...         1       0  698318dd7
2829    yes     no   72    ...         0       1  4559a5af0
2830    yes     no   76    ...         1       0  4559a5af0
2831    yes     no   49    ...         0       1  4559a5af0
3015    yes     no   10    ...         1       0  78f517dc7
3016    yes     no   35    ...         1       0  78f517dc7
3017    yes     no   73    ...         1       0  78f517dc7
3741    yes     no    0    ...         1       0  0b3d5369b
3742    yes     no    2    ...         0       1  0b3d5369b
3743    yes     no    6    ...         1       0  0b3d5369b
3744    yes     no   31    ...         1       0  0b3d5369b
3745    yes     no   29    ...         0       1  0b3d5369b
4490    yes     no   56    ...         0       1  f2ed4871d
4491    yes     no   67    ...         1       0  f2ed4871d
4492    yes     no   16    ...         0       1  f2ed4871d
4493    yes     no    5    ...         0       1  f2ed4871d
4494    yes     no   28    ...         0       1  f2ed4871d
4903    yes     no   22    ...         1       0  e26adc10d
4904    yes     no   50    ...         0       1  e26adc10d
4905    yes     no   59    ...         1       0  e26adc10d
4979    yes     no   17    ...         1       0  0519a23f5
4980    yes     no   46    ...         1       0  0519a23f5
4981    yes     no   52    ...         0       1  0519a23f5
...     ...    ...  ...    ...       ...     ...        ...
8385    yes     no   33    ...         1       0  41f4d7a4b
8585    yes     no   78    ...         1       0  12a23bffe
8903    yes     no   30    ...         1       0  180d40332
8904    yes     no   24    ...         0       1  180d40332
8905    yes     no    7    ...         1       0  180d40332
8951    yes     no   13    ...         0       1  d4bce9879
8952    yes     no   48    ...         0       1  d4bce9879
8953    yes     no   17    ...         0       1  d4bce9879
8954    yes     no   60    ...         1       0  d4bce9879
8955    yes     no   28    ...         1       0  d4bce9879
8956    yes     no    1    ...         0       1  d4bce9879
8957    yes     no   26    ...         0       1  d4bce9879
9040    yes     no   30    ...         0       1  da2ecdcfd
9041    yes     no   13    ...         1       0  da2ecdcfd
9042    yes     no    9    ...         1       0  da2ecdcfd
9043    yes     no   15    ...         1       0  da2ecdcfd
9044    yes     no   23    ...         1       0  da2ecdcfd
9045    yes     no    1    ...         1       0  da2ecdcfd
9046    yes     no   11    ...         0       1  da2ecdcfd
9047    yes     no    4    ...         0       1  da2ecdcfd
9197    yes     no   43    ...         0       1  5fd40ddd3
9198    yes     no   23    ...         1       0  5fd40ddd3
9199    yes     no   60    ...         1       0  5fd40ddd3
9257    yes     no   75    ...         1       0  7e0dc2f87
9415    yes     no   73    ...         0       1  7dde74368
9416    yes     no   47    ...         1       0  7dde74368
9417    yes     no   82    ...         1       0  7dde74368
9463    yes     no   68    ...         1       0  be108a783
9464    yes     no   57    ...         0       1  be108a783
9465    yes     no   35    ...         1       0  be108a783

[123 rows x 8 columns]

In [18]:
train[(train['edjefe']=='no') & (train['edjefa']=='no')][['edjefe', 'edjefa', 'age', 'escolari', 'female', 'male', 'Id', 'parentesco1','idhogar']]

edjefe edjefa  age    ...                Id  parentesco1    idhogar
86       no     no   89    ...      ID_48c10ba80            1  2b1a06ddc
87       no     no   55    ...      ID_1a283d51b            0  2b1a06ddc
121      no     no   21    ...      ID_fdf18bbee            0  d9e9b591a
122      no     no   47    ...      ID_cba836ed8            0  d9e9b591a
123      no     no   49    ...      ID_6031cb88d            0  d9e9b591a
124      no     no   68    ...      ID_0bc817cb2            1  d9e9b591a
125      no     no   21    ...      ID_1cfb76926            0  d9e9b591a
166      no     no    6    ...      ID_f435c5e46            0  71bec40bc
167      no     no    9    ...      ID_a51262dae            0  71bec40bc
168      no     no   26    ...      ID_7e6d23b28            0  71bec40bc
169      no     no   47    ...      ID_4c409e16e            0  71bec40bc
170      no     no    4    ...      ID_0704f4d9f            0  71bec40bc
171      no     no   60    ...      ID_b0adca84e            1  71bec40bc
172      no     no   25    ...      ID_0e4c312aa            0  71bec40bc
416      no     no   34    ...      ID_2399e1e3e            1  fd22b4607
417      no     no   34    ...      ID_301d72522            0  fd22b4607
418      no     no    7    ...      ID_d643847f2            0  fd22b4607
419      no     no   14    ...      ID_02a405b3b            0  fd22b4607
498      no     no   36    ...      ID_460e4f901            1  7bf86807a
499      no     no    4    ...      ID_0701248a7            0  7bf86807a
500      no     no   11    ...      ID_a643ace5b            0  7bf86807a
501      no     no   35    ...      ID_fcdff98ad            0  7bf86807a
502      no     no   17    ...      ID_3a60d3642            0  7bf86807a
634      no     no   21    ...      ID_01f647aaf            0  fd10905bc
635      no     no   85    ...      ID_6a13e748d            1  fd10905bc
636      no     no    8    ...      ID_e7fa38760            0  fd10905bc
637      no     no   36    ...      ID_0d3b0550d            0  fd10905bc
669      no     no    0    ...      ID_76517e7d8            0  e8725f7cb
670      no     no   73    ...      ID_c30b3f5c4            0  e8725f7cb
671      no     no   68    ...      ID_a5bedc6c4            1  e8725f7cb
...     ...    ...  ...    ...               ...          ...        ...
9342     no     no   73    ...      ID_b75c1d362            1  152d719e2
9343     no     no   72    ...      ID_01462b000            0  152d719e2
9374     no     no   20    ...      ID_b4f231ccb            0  7b7ebaf70
9375     no     no    5    ...      ID_46eb2842b            0  7b7ebaf70
9376     no     no   46    ...      ID_d7f45a459            0  7b7ebaf70
9377     no     no   25    ...      ID_5b5fbe311            0  7b7ebaf70
9378     no     no   26    ...      ID_aed542da7            0  7b7ebaf70
9379     no     no    8    ...      ID_12c10b214            0  7b7ebaf70
9380     no     no   15    ...      ID_cd64c1af9            0  7b7ebaf70
9381     no     no   52    ...      ID_d43b03397            1  7b7ebaf70
9389     no     no   97    ...      ID_06caac58e            0  b4bfd5115
9390     no     no   14    ...      ID_635e2f20e            0  b4bfd5115
9391     no     no   68    ...      ID_2129e7f05            1  b4bfd5115
9431     no     no   21    ...      ID_65d174f99            0  0966c7521
9432     no     no   70    ...      ID_3ef316d48            1  0966c7521
9433     no     no   57    ...      ID_317e9a221            0  0966c7521
9434     no     no   17    ...      ID_adfb4cf14            0  0966c7521
9450     no     no   40    ...      ID_f0886bae3            0  2ca73280c
9451     no     no   69    ...      ID_3f66348ab            1  2ca73280c
9452     no     no   33    ...      ID_c51d7c66b            0  2ca73280c
9453     no     no    8    ...      ID_2f5ad837a            0  2ca73280c
9454     no     no   41    ...      ID_d74271567            0  2ca73280c
9455     no     no   59    ...      ID_dc7068500            0  

In [19]:
train[(train['edjefe']=='yes') & train['parentesco1']==1][['escolari']]

escolari
601          1
2394         1
2424         1
2830         1
3017         1
3744         1
4491         1
4905         1
4980         1
5154         1
5191         1
5296         1
5338         1
5387         1
5562         1
5579         1
5588         1
5728         1
6342         1
6605         1
6665         1
6743         1
6869         1
6884         1
7056         1
7079         1
7361         1
7368         1
7759         1
7833         1
7907         1
8384         1
8585         1
8903         1
8954         1
9044         1
9199         1
9257         1
9417         1
9463         1

In [20]:
conditions = [
    (train['edjefe']=='no') & (train['edjefa']=='no'), #both no
    (train['edjefe']=='yes') & (train['edjefa']=='no'), # yes and no
    (train['edjefe']=='no') & (train['edjefa']=='yes'), #no and yes 
    (train['edjefe']!='no') & (train['edjefe']!='yes') & (train['edjefa']=='no'), # number and no
    (train['edjefe']=='no') & (train['edjefa']!='no') # no and number
]
choices = [0, 1, 1, train['edjefe'], train['edjefa']]
train['edjefx']=np.select(conditions, choices)
train['edjefx']=train['edjefx'].astype(int)
train[['edjefe', 'edjefa', 'edjefx']][:15]

edjefe edjefa  edjefx
0      10     no      10
1      12     no      12
2      no     11      11
3      11     no      11
4      11     no      11
5      11     no      11
6      11     no      11
7       9     no       9
8       9     no       9
9       9     no       9
10      9     no       9
11     no     11      11
12     no     11      11
13     no      4       4
14     no      4       4

In [21]:
train['edjefe'].dtypes

dtype('O')

In [22]:
train.columns[train.isna().sum()!=0]

Index(['v2a1', 'v18q1', 'rez_esc', 'meaneduc', 'SQBmeaned'], dtype='object')

Columns with nans:

    v2a1 - monthly rent
    v18q1 - number of tablets
    rez_esc - years behind school
    meaneduc - mean education for adults
    SQBmeaned - square of meaned

In [23]:
train[train['meaneduc'].isnull()]

Id      v2a1  hacdor   ...    agesq  Target  edjefx
1291  ID_bd8e11b0f       NaN       0   ...      324       4      10
1840  ID_46ff87316  110000.0       0   ...      324       4       4
1841  ID_69f50bf3e  110000.0       0   ...      324       4       4
2049  ID_db3168f9f  180000.0       0   ...      361       4      12
2050  ID_2a7615902  180000.0       0   ...      361       4      12

[5 rows x 144 columns]

In [24]:
print(len(train[train['idhogar']==train.iloc[1291]['idhogar']]))
print(len(train[train['idhogar']==train.iloc[1840]['idhogar']]))
print(len(train[train['idhogar']==train.iloc[2049]['idhogar']]))

1
2
2


So, the 5 rows with Nan for 'meaneduc' is just 3 households, where 18-19 year-olds live. No other people live in these households. Then we can just take the education levels of these kids ('escolari') and put them into 'meaneduc' and 'SQBmeaned'.

In [25]:
meaneduc_nan=train[train['meaneduc'].isnull()][['Id','idhogar','escolari']]

In [26]:
me=meaneduc_nan.groupby('idhogar')['escolari'].mean().reset_index()

In [27]:
me

idhogar  escolari
0  1b31fd159        10
1  a874b7ce7         5
2  faaebf71a        12

In [28]:
for row in meaneduc_nan.iterrows():
    idx=row[0]
    idhogar=row[1]['idhogar']
    m=me[me['idhogar']==idhogar]['escolari'].tolist()[0]
    train.at[idx, 'meaneduc']=m
    train.at[idx, 'SQBmeaned']=m*m

Next, let's look at 'v2a1', the monthly rent payment, that also has missing values.


In [29]:
train['v2a1'].isnull().sum()

6860

That's a lot of missing values.

But maybe they own their houses. We can look what type of housing these people with missing values live..


In [30]:
norent=train[train['v2a1'].isnull()]
print("Owns his house:", norent[norent['tipovivi1']==1]['Id'].count())
print("Owns his house paying installments", norent[norent['tipovivi2']==1]['Id'].count())
print("Rented ", norent[norent['tipovivi3']==1]['Id'].count())
print("Precarious ", norent[norent['tipovivi4']==1]['Id'].count())
print("Other ", norent[norent['tipovivi5']==1]['Id'].count())
print("Total ", 6860)

Owns his house: 5911
Owns his house paying installments 0
Rented  0
Precarious  163
Other  786
Total  6860


The majority in fact owns their houses, only a few have odd situations. We can probably just assume they don't pay rent, and put 0 in these cases.

In [31]:
train['v2a1']=train['v2a1'].fillna(0)

Now, let's look at 'v18q1', which indicates how many tablets the household owns

In [32]:
train['v18q1'].isna().sum()

7342

In [33]:
train['v18q1'].unique()

array([nan,  1.,  2.,  3.,  4.,  5.,  6.])

That's exactly the number of rows with missing values! There's also no 0 among the values of 'v18q1'. So all the nans in 'v18q1' just means they don't own a tablet! So we can just change them to 0.

In [34]:
train['v18q1']=train['v18q1'].fillna(0)

Next up is 'rez_esc', which indicates if a person is behind in school.


In [35]:
train['rez_esc'].isnull().sum()

7928

In [36]:
train['rez_esc'].describe()

count    1629.000000
mean        0.459791
std         0.946550
min         0.000000
25%         0.000000
50%         0.000000
75%         1.000000
max         5.000000
Name: rez_esc, dtype: float64

In [37]:
train['rez_esc'].unique()

array([nan,  1.,  0.,  3.,  2.,  4.,  5.])

In [38]:
train[train['rez_esc']>1][['age', 'escolari', 'rez_esc']][:20]

age  escolari  rez_esc
185    13         3      3.0
190    16         7      2.0
240    16         6      3.0
242    11         2      2.0
312    15         6      2.0
317    17         6      4.0
319    17         7      3.0
320    15         6      2.0
333    16         7      2.0
510    16         7      2.0
617    16         6      3.0
629    16         6      3.0
714    17         6      4.0
769    17         6      4.0
837    16         6      3.0
937    16         6      3.0
978    16         7      2.0
1146   16         6      3.0
1150   15         6      2.0
1151   16         7      2.0

these are all schoolchildren...


In [39]:
rez_esc_nan=train[train['rez_esc'].isnull()]
rez_esc_nan[(rez_esc_nan['age']<18) & rez_esc_nan['escolari']>0][['age', 'escolari']]

Empty DataFrame
Columns: [age, escolari]
Index: []

So all the nans here are either adults or children before school age. We can input 0 again.

In [40]:
train['rez_esc']=train['rez_esc'].fillna(0)


same household can have different target values

In [41]:
d={}
weird=[]
for row in train.iterrows():
    idhogar=row[1]['idhogar']
    target=row[1]['Target']
    if idhogar in d:
        if d[idhogar]!=target:
            weird.append(idhogar)
    else:
        d[idhogar]=target

In [42]:
len(set(weird))

85

There are 85 households like that.


In [43]:
train[train['idhogar']==weird[2]][['idhogar','parentesco1', 'Target']]

idhogar  parentesco1  Target
285  6833ac5dc            0       2
286  6833ac5dc            0       2
287  6833ac5dc            0       2
288  6833ac5dc            1       2
289  6833ac5dc            0       2
290  6833ac5dc            0       1

Correct target value is the one belonging to the head of the household. So we should set the correct value each time.

In [44]:
for i in set(weird):
    hhold=train[train['idhogar']==i][['idhogar', 'parentesco1', 'Target']]
    target=hhold[hhold['parentesco1']==1]['Target'].tolist()[0]
    for row in hhold.iterrows():
        idx=row[0]
        if row[1]['parentesco1']!=1:
            train.at[idx, 'Target']=target

In [45]:
train[train['idhogar']==weird[1]][['idhogar','parentesco1', 'Target']]

idhogar  parentesco1  Target
282  4b6077882            1       1
283  4b6077882            0       1
284  4b6077882            0       1

In [46]:
def data_cleaning(train):
    train['dependency']=np.sqrt(train['SQBdependency'])
    train['rez_esc']=train['rez_esc'].fillna(0)
    train['v18q1']=train['v18q1'].fillna(0)
    train['v2a1']=train['v2a1'].fillna(0)
    
    conditions = [
    (str(train['edjefe'])=='no') & (str(train['edjefa'])=='no'), #both no
    (str(train['edjefe'])=='yes') & (str(train['edjefa'])=='no'), # yes and no
    (str(train['edjefe'])=='no') & (str(train['edjefa'])=='yes'), #no and yes 
    (str(train['edjefe'])!='no') & (str(train['edjefe'])!='yes') & (str(train['edjefa'])=='no'), # number and no
    (str(train['edjefe'])=='no') & (str(train['edjefa'])!='no') # no and number
    ]
    choices = [0, 1, 1, train['edjefe'], train['edjefa']]
    train['edjefx']=np.select(conditions, choices)
    train['edjefx']=train['edjefx'].astype(int)
    ##train.drop(['edjefe', 'edjefa'], axis=1, inplace=True)
    
    meaneduc_nan=train[train['meaneduc'].isnull()][['Id','idhogar','escolari']]
    me=meaneduc_nan.groupby('idhogar')['escolari'].mean().reset_index()
    for row in meaneduc_nan.iterrows():
        idx=row[0]
        idhogar=row[1]['idhogar']
        m=me[me['idhogar']==idhogar]['escolari'].tolist()[0]
        train.at[idx, 'meaneduc']=m
        train.at[idx, 'SQBmeaned']=m*m
        
    return train

In [47]:
train.head()

Id      v2a1  hacdor   ...    agesq  Target  edjefx
0  ID_279628684  190000.0       0   ...     1849       4      10
1  ID_f29eb3ddd  135000.0       0   ...     4489       4      12
2  ID_68de51c94       0.0       0   ...     8464       4      11
3  ID_d671db89c  180000.0       0   ...      289       4      11
4  ID_d56d6f5f5  180000.0       0   ...     1369       4      11

[5 rows x 144 columns]

In [48]:
train['edjefe'].dtypes

dtype('O')

In [49]:
target = train['Target'].value_counts().to_frame()
levels = ["NonVulnerable", "Moderate Poverty", "Vulnerable", "Extereme Poverty"]
trace = go.Bar(y=target.Target, x=levels, marker=dict(color='orange', opacity=0.6))
layout = dict(title="Household Poverty Levels", margin=dict(l=200), width=800, height=400)
data1 = [trace]
fig = go.Figure(data=data1, layout=layout)
iplot(fig)

The above graph shows the poverty level distribution. While the majority households lie in Non vulnerable level, a significant households comes in vulnerable and extreme poverty condition.


# Modeling

## Light Gradient Boosting Machine

### Baseline Model

In [50]:
features = [c for c in train.columns if c not in ['Id', 'Target']]

In [51]:
target = train['Target'].values
target_index = {1:0, 2:1, 3:2, 4:3}
target = np.array([target_index[c] for c in target])

In [52]:
def label_encoding(col):
    le = LabelEncoder()
    le.fit(list(train[col].values) + list(test[col].values))
    train[col] = le.transform(train[col].astype(str))
    test[col] = le.transform(test[col].astype(str))

num_cols = train._get_numeric_data().columns
cat_cols = list(set(features) - set(num_cols))
for col in cat_cols:
    label_encoding(col)

### Prepare Training and Validation Sets


In [53]:
X_train, X_valid, y_train, y_valid = train_test_split(train[features].values, target, test_size=0.2, random_state=1)
lgb_train = lgb.Dataset(X_train, y_train)
lgb_valid = lgb.Dataset(X_valid, y_valid)

In [54]:
params = {'boosting_type': 'gbdt', 'objective': 'multiclass', 'metric': 'multi_logloss',
          'num_class': 4, 'max_depth': 44, 'num_leaves': 36, 'learning_rate': 0.01,
          'feature_fraction': 0.8, 'bagging_fraction': 0.8, 'bagging_freq': 5,
          'lambda_l2': 1.0, 'verbose': -1, 'num_threads': -1 }

model = lgb.train(params, lgb_train, num_boost_round=100,  valid_sets=[lgb_train, lgb_valid], 
                  early_stopping_rounds=2000, verbose_eval=100)

Training until validation scores don't improve for 2000 rounds.
[100]	training's multi_logloss: 0.727722	valid_1's multi_logloss: 0.771635
Did not meet early stopping. Best iteration is:
[100]	training's multi_logloss: 0.727722	valid_1's multi_logloss: 0.771635


### Predict and Submit


In [55]:
test_data=data_cleaning(test)
features1 = [c for c in train.columns if c not in ['Id', 'Target']]

## predictions
preds = model.predict(test_data[features1], num_iteration=model.best_iteration)

## revert back to original classes
reverse_index = {}
for k,v in target_index.items():
    reverse_index[v] = k
preds = [reverse_index[np.argmax(p)] for p in preds]

## submission
subs = pd.DataFrame()
subs['Id'] = test_data['Id']
subs['Target'] = preds
subs.to_csv('submission_LGBM.csv', index=False)
subs.head()

Id  Target
0  ID_2f6873615       4
1  ID_1c78846d2       4
2  ID_e5442cf6a       4
3  ID_a8db26a79       4
4  ID_a62966799       4

## Random Forest

In [56]:
data_undersampled=train.drop(train.query('Target == 4').sample(frac=.75).index)

In [57]:
X=data_undersampled.drop(['Id', 'idhogar', 'Target', 'edjefe', 'edjefa'], axis=1)
y=data_undersampled['Target']

In [58]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

In [59]:
X_train.shape

(4043, 139)

In [60]:
y_train.shape

(4043,)

In [61]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import GridSearchCV

In [62]:
clf = RandomForestClassifier()
params={'n_estimators': list(range(40,61, 1))}
gs = GridSearchCV(clf, params, cv=5)

In [63]:
gs.fit(X_train, y_train)

GridSearchCV(cv=5, error_score='raise-deprecating',
       estimator=RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators='warn', n_jobs=None,
            oob_score=False, random_state=None, verbose=0,
            warm_start=False),
       fit_params=None, iid='warn', n_jobs=None,
       param_grid={'n_estimators': [40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring=None, verbose=0)

In [64]:
preds=gs.predict(X_test)

In [65]:
from sklearn.metrics import classification_report
print(classification_report(y_test, preds))

              precision    recall  f1-score   support

           1       0.94      0.93      0.93       154
           2       0.86      0.92      0.89       317
           3       0.89      0.85      0.87       236
           4       0.86      0.84      0.85       304

   micro avg       0.88      0.88      0.88      1011
   macro avg       0.89      0.88      0.89      1011
weighted avg       0.88      0.88      0.88      1011



In [66]:
from sklearn.metrics import confusion_matrix
print(confusion_matrix(y_test, preds))

[[143   2   0   9]
 [  2 291   9  15]
 [  3  16 200  17]
 [  4  28  16 256]]


In [67]:
print(gs.best_params_)
print(gs.best_score_)
print(gs.best_estimator_)

{'n_estimators': 51}
0.8684145436557013
RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=51, n_jobs=None,
            oob_score=False, random_state=None, verbose=0,
            warm_start=False)


In [68]:
cvres = gs.cv_results_
for mean_score, params in zip(cvres["mean_test_score"], cvres["params"]):
    print(np.sqrt(mean_score), params)

0.9298948474991094 {'n_estimators': 40}
0.928697129130486 {'n_estimators': 41}
0.9281643137514988 {'n_estimators': 42}
0.9266974918520592 {'n_estimators': 43}
0.9253619997446177 {'n_estimators': 44}
0.9262970463921579 {'n_estimators': 45}
0.9258964277416636 {'n_estimators': 46}
0.9290965401427713 {'n_estimators': 47}
0.9306924701025122 {'n_estimators': 48}
0.9274978640972825 {'n_estimators': 49}
0.9308253407535002 {'n_estimators': 50}
0.9318876239416968 {'n_estimators': 51}
0.9304266718762999 {'n_estimators': 52}
0.9308253407535002 {'n_estimators': 53}
0.9292296389959924 {'n_estimators': 54}
0.9276311923325858 {'n_estimators': 55}
0.9294957795252919 {'n_estimators': 56}
0.9306924701025122 {'n_estimators': 57}
0.9310910251718559 {'n_estimators': 58}
0.9314894097120157 {'n_estimators': 59}
0.9294957795252919 {'n_estimators': 60}


In [69]:
test_data1=pd.read_csv('../input/test.csv')

In [70]:
test_data1.head()

Id      v2a1  hacdor  ...    SQBdependency  SQBmeaned  agesq
0  ID_2f6873615       NaN       0  ...             0.25     272.25     16
1  ID_1c78846d2       NaN       0  ...             0.25     272.25   1681
2  ID_e5442cf6a       NaN       0  ...             0.25     272.25   1681
3  ID_a8db26a79       NaN       0  ...             0.00     256.00   3481
4  ID_a62966799  175000.0       0  ...            64.00        NaN    324

[5 rows x 142 columns]

In [71]:
test_data2=data_cleaning(test_data1)
test_data2.drop(['edjefe','edjefa'],axis=1,inplace=True)

In [72]:
ids=test_data2['Id']
test_data2.drop(['Id', 'idhogar'], axis=1, inplace=True)

In [73]:
test_predictions=gs.predict(test_data2)

In [74]:
test_predictions[:5]

array([4, 4, 4, 4, 4])

In [75]:
submit=pd.DataFrame({'Id': ids, 'Target': test_predictions})

In [76]:
submit.to_csv('submit.csv', index=False)

# Conclusion


This study has revealed the level of Poverty for Costa Rican families and classified them into 4 types namely Vulnearble, Not Vulnerable, Moderate and Extremely Vulnerable. they have been numbered from 1 to 4 respectively. The classification has been done with an F1 score of 0.22 by LightGBM method and 0.3760.

In the process complete Exploratory data analysis has been done to analyze various factors affecting the poverty level of families. Some of the prominent factors were schooling, condition of houses, education of head of family, assets that households have like TV, refrigerator, tablets etc. All these were taken into account while analyzing the poverty levels.

Two algorithms were implemented namely LightGBM and Random Forest. Because of low Macro F score with LightGBM, Random Forest has been preferred which has given a score of 0.41064.

# Contribution

By Own: 30%
Created code for data cleaning and modified LightGBM code for our implementation

By External Source: 70%

# Citations

https://www.kaggle.com/mandan/lightgbm-benchmark-implementation

https://www.kaggle.com/c/costa-rican-household-poverty-prediction/data

https://www.kaggle.com/katacs/data-cleaning-and-random-forest

# Licence 


Copyright 2019 Ashu Kapil

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.